xl2times/transforms.py

```diff
@@ -2227,78 +2321,19 @@ def apply_transform_tables(
     Include data from transformation tables.
     """
 
-    topology = generate_topology_dictionary(tables, model)
-
-    def query(
-        table: DataFrame,
-        process: str | None,
-        commodity: str | None,
-        attribute: str | None,
-        region: str | None,
-        year: int | None,
-    ) -> pd.Index:
-        qs = []
-        if process is not None:
-            qs.append(f"process in ['{process}']")
-        if commodity is not None:
-            qs.append(f"commodity in ['{commodity}']")
-        if attribute is not None:
-            qs.append(f"attribute == '{attribute}'")
-        if region is not None:
-            qs.append(f"region == '{region}'")
-        if year is not None:
-            qs.append(f"year == {year}")
-        return table.query(" and ".join(qs)).index
-
-    def eval_and_update(
-        table: DataFrame, rows_to_update: pd.Index, new_value: str
-    ) -> None:
-        """Performs an inplace update of rows `rows_to_update` of `table` with `new_value`,
-        which can be a update formula like `*2.3`."""
-        if isinstance(new_value, str) and new_value[0] in {"*", "+", "-", "/"}:
-            old_values = table.loc[rows_to_update, "value"]
-            updated = old_values.astype(float).map(lambda x: eval("x" + new_value))
-            table.loc[rows_to_update, "value"] = updated
-        else:
-            table.loc[rows_to_update, "value"] = new_value
-
     if datatypes.Tag.tfm_upd in tables:
         updates = tables[datatypes.Tag.tfm_upd]
         table = tables[datatypes.Tag.fi_t]
-        new_tables = [table]
+
         # Reset FI_T index so that queries can determine unique rows to update
         tables[datatypes.Tag.fi_t].reset_index(inplace=True, drop=True)
 
-        # TFM_UPD: expand wildcards in each row, query FI_T to find matching rows,
-        # evaluate the update formula, and add new rows to FI_T
-        # TODO perf: collect all updates and go through FI_T only once?
-        for _, row in tqdm(
-            updates.iterrows(),
-            total=len(updates),
-            desc=f"Applying transformations from {datatypes.Tag.tfm_upd.value}",
-        ):
-            rows_to_update = query(
-                table,
-                row["process"],
-                row["commodity"],
-                row["attribute"],
-                row["region"],
-                row["year"],
-            )
+        # Apply the update formulas in updates to the rows in table
+        new_tables = _eval_updates(table, updates)
 
-            if not any(rows_to_update):
-                logger.info(
-                    f"A {datatypes.Tag.tfm_upd.value} row generated no records."
-                )
-                continue
-
```

Comment:
I think we should have a warning when a row doesn't generate a record. I believe your changes from #210 could reduce the number of warnings by e. g. passing a list of processes generated by a wildcard to query (currently they are passed individually)

Reply:
Hmm... I was getting thousands of these log messages before, so I'm not sure how useful this actually was.  

I'll have a think about how to identify the `update`s that don't match anything in the `table` though, as maybe a single summary would be better here, like: `5000 update rows from the 'tfm_upd' table matched no rows in 'fi_t'`

Reply:
> I was getting thousands of these log messages before, so I'm not sure how useful this actually was.

It is because currently a single wildcard results in multiple update rows. I did not do anything about it because I did not want to replicate the changes you were doing in #210 (e.g. not exploding the lists of processes/commodities). It would be really great if ` _eval_updates` could handle lists of processes/commodities as `query` does.

Reply:
Cool, I think I've worked out a way to query for the non-matches.
Yeah I did wonder if we could do the evals and stuff here with non-exploded process/commodity columns though.  That'd certainly make it a lot faster....